clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -386,15 +441,42 @@ func pivot(from, to ClusterClient) error {
 	for _, machine := range machines {
 		// New objects cannot have a specified resource version. Clear it out.
 		machine.SetResourceVersion("")
-		err = to.CreateMachineObjects([]*clusterv1.Machine{machine})
-		if err != nil {
-			return err
+		if err = to.CreateMachineObjects([]*clusterv1.Machine{machine}); err != nil {
+			return fmt.Errorf("error moving Machine '%v': %v", machine.GetName(), err)
 		}
 		glog.Infof("Moved Machine '%s'", machine.GetName())
 	}
 	return nil
 }
 
+func deleteObjects(client ClusterClient) error {
+	var errors []string
+	glog.Infof("Deleting machine deployments")
+	if err := client.DeleteMachineDeploymentObjects(); err != nil {
+		err = fmt.Errorf("error deleting machine deployments: %v", err)
+		errors = append(errors, err.Error())
+	}
+	glog.Infof("Deleting machine sets")
+	if err := client.DeleteMachineSetObjects(); err != nil {
+		err = fmt.Errorf("error deleting machine sets: %v", err)
+		errors = append(errors, err.Error())
+	}
+	glog.Infof("Deleting machines")
+	if err := client.DeleteMachineObjects(); err != nil {
```

Comment:
It's safe in the sense that we are simply calling the 'delete all' method in the cluster API. It is up to the cluster API to properly handle machines that already have a deletion in progress due to the fact that their parent object has been deleted.

Reply:
It's still not great because attempting to delete a machine twice (or three times) could spit a couple of error logs which would be misleading unless the user knows exactly how this code works, and exactly how MachineSets work. I think here, we should delete all Machines that are not in a MachineSet.

Reply:
There is no actual deletion going on here, there is just a call to the cluster-api "delete all" methods, i.e. delete all machines, delete all machinesets, delete all machine deployments. Those are asynchronous methods that are eventually reconciled by the controllers. IN this case, the underlying deletion for machines & machine sets is done by the machine controller. 

I'm not convinced there actually is a problem or that the cluster API would do the wrong thing. Rather than put a lot of complicated deletion logic in clusterctl it seems better to have the controllers do the right thing.

Reply:
I synced with @k4leung4  and he confirmed that there is not an issue here, there are a couple scenarios / points that we discussed:

1. **Deleting existing Machines?**: DeleteCollection(...) (the actual delete methods being used) do not return errors when there are already objects that are in the process of being deleted (i.e. marked as being deleted). We are not passing in any machines we are simply call ing a "delete all" method.
2. **Concurrent Deletes?**: We are calling the DeleteCollection(...) method and passing in the propogation policy of DeletePropogationForeground. This means the delete call will block until all child objects (i.e. machines that belong to a machine set are deleted.
3. **Deleting Machines before MachineSets**: Even if we did things in the wrong order and called DeleteCollection on Machines, and that method actually deleted machines associated with a MachineSet (I've not yet drilled into whether that would even happen) all that would occur is the controller would attempt to recreate the machine and then when the DeleteCollection was called on the machine sets the machines would be deleted again.